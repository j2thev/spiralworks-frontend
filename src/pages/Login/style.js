import styled from 'styled-components';
import { Container } from 'reactstrap'; 

export const LoginContainer = styled(Container)`
  margin-top: 50px;
`;
