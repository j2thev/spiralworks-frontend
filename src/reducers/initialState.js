export default {
  user: {
    id: '',
    email: '',
    phoneNo: '',
    firstName: '',
    lastName: '',
    isAuthenticated: false
  }
}
