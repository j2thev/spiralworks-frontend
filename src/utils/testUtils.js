import React from 'react';
import { Provider } from 'react-redux';

export function TestProvider({ children, store }) {
  return <Provider store={store}>{children}</Provider>;
}
