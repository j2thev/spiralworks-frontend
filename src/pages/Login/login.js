import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  Row,
  Col,
  Form,
  FormGroup,
  Label,
  Input,
  Button,
  FormText,
  Alert
} from 'reactstrap';
import * as L from './style';
import * as userActions from '../../actions/userActions';
import { map, isEmpty } from 'lodash';
import history from '../../utils/history';

class Login extends Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);

    this.state = {
      username: '',
      password: '',
      errors: {
        username: '',
        password: ''
      },
      alert: {
        color: '',
        message: '',
        visible: false
      }
    }
  }

  handleChange(event) {
    const { name, value } = event.target;
    const { errors } = this.state;

    switch(name) {
      case 'username':
      case 'password': {
        Object.assign(errors, { [name]: '' });
        break;
      }
      default: {
        break;
      }
    }

    this.setState({
      [name]: value,
      errors
    }, () => console.log(this.state));
  }

  onShowAlert(alert) {
    this.setState({ alert }, () => {
      setTimeout(() => {
        this.setState({
          alert: {
            visible: false
          }
        });
      }, 5000);
    });
  }

  validateForm(errors) {
    let valid = true;

    map(errors, (error) => error.length > 0 && (valid = false));

    return valid;
  }

  handleSubmit(event) {
    event.preventDefault();

    const { login, update } = this.props;
    const { username, password, errors } = this.state;
    const user = { username, password };

    map(event.target, (node) => {
      const { name, validationMessage } = node;

      Object.assign(errors, { [name]: validationMessage });
    });

    if (this.validateForm(errors)) {
      login(user)
        .then(result => {
          const { data } = result.data;

          if (!isEmpty(data)) {
            const { _id: id, email, phoneNo, firstName, lastName } = data;
            const user = {
              id,
              email,
              phoneNo,
              firstName,
              lastName,
              isAuthenticated: true
            };

            update(user);

            history.push('/profile');
          } else {
            const alert = {
              color: 'danger',
              message: 'Incorrect username or password',
              visible: true
            };

            this.onShowAlert(alert);
          }
        })
        .catch(error => {
          const { message } = error;
          const alert = {
            color: 'danger',
            message,
            visible: true
          };

          this.onShowAlert(alert);
        });
    } else {
      this.setState({ errors });
    }

  }

  render() {
    const { username, password, errors, alert } = this.state;

    return (
      <L.LoginContainer>
        <Row>
          <Col sm="12" md={{ size: 8, offset: 2 }}>
            <Form onSubmit={this.handleSubmit} noValidate>
              <FormGroup>
                <h3>Sign in</h3>
              </FormGroup>
              <FormGroup>
                <Alert color={alert.color} isOpen={alert.visible}>
                  {alert.message}
                </Alert>
              </FormGroup>
              <FormGroup>
                <Label>Email or Phone Number</Label>
                <Input
                  invalid={errors.username.length > 0 ? true : false}
                  type="text"
                  required
                  name="username"
                  value={username}
                  autoFocus
                  onChange={this.handleChange}
                />
                <FormText color="danger">
                  {errors.username}
                </FormText>
              </FormGroup>
              <FormGroup>
                <Label>Password</Label>
                <Input
                  invalid={errors.password.length > 0 ? true : false}
                  type="password"
                  required
                  name="password"
                  value={password}
                  onChange={this.handleChange}
                />
                <FormText color="danger">
                  {errors.password}
                </FormText>
              </FormGroup>
              <Button block color="primary">SIGN IN</Button>
            </Form>
          </Col>
        </Row>
      </L.LoginContainer>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(userActions, dispatch);
};

export default connect(null, mapDispatchToProps)(Login);
