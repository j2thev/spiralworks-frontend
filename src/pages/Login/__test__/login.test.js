import React from 'react';
import configuredMockStore from 'redux-mock-store';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import { TestProvider } from '../../../utils/testUtils';
import initialState from '../../../reducers/initialState';
import '../../../tools/setupEnzyme';
import Login from '../../Login';

describe('Login Page', () => {
  it('should render correctly', () => {
    const store = configuredMockStore()(initialState);
    const container = mount(
      <TestProvider store={store}>
        <Login />
      </TestProvider>
    );

    expect(toJSON(container)).toMatchSnapshot();
  });
});
