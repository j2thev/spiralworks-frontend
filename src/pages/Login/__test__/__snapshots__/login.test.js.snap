// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render correctly 1`] = `
<TestProvider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Connect(Login)>
      <Login
        create={[Function]}
        login={[Function]}
        logout={[Function]}
        save={[Function]}
        update={[Function]}
      >
        <Styled(Container)>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bdVaJa",
                  "isStatic": true,
                  "lastClassName": "bxvob",
                  "rules": Array [
                    "
  margin-top: 50px;
",
                  ],
                },
                "displayName": "Styled(Container)",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bdVaJa",
                "target": [Function],
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <Container
              className="sc-bdVaJa bxvob"
              tag="div"
            >
              <div
                className="sc-bdVaJa bxvob container"
              >
                <Row
                  tag="div"
                >
                  <div
                    className="row"
                  >
                    <Col
                      md={
                        Object {
                          "offset": 2,
                          "size": 8,
                        }
                      }
                      sm="12"
                      tag="div"
                      widths={
                        Array [
                          "xs",
                          "sm",
                          "md",
                          "lg",
                          "xl",
                        ]
                      }
                    >
                      <div
                        className="col-sm-12 col-md-8 offset-md-2"
                      >
                        <Form
                          noValidate={true}
                          onSubmit={[Function]}
                          tag="form"
                        >
                          <form
                            className=""
                            noValidate={true}
                            onSubmit={[Function]}
                          >
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <h3>
                                  Sign in
                                </h3>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Alert
                                  closeAriaLabel="Close"
                                  color=""
                                  fade={true}
                                  isOpen={false}
                                  tag="div"
                                  transition={
                                    Object {
                                      "appear": true,
                                      "baseClass": "fade",
                                      "baseClassActive": "show",
                                      "enter": true,
                                      "exit": true,
                                      "in": true,
                                      "mountOnEnter": false,
                                      "onEnter": [Function],
                                      "onEntered": [Function],
                                      "onEntering": [Function],
                                      "onExit": [Function],
                                      "onExited": [Function],
                                      "onExiting": [Function],
                                      "tag": "div",
                                      "timeout": 150,
                                      "unmountOnExit": true,
                                    }
                                  }
                                >
                                  <Fade
                                    appear={true}
                                    baseClass="fade"
                                    baseClassActive="show"
                                    className="alert alert-"
                                    enter={true}
                                    exit={true}
                                    in={false}
                                    mountOnEnter={false}
                                    onEnter={[Function]}
                                    onEntered={[Function]}
                                    onEntering={[Function]}
                                    onExit={[Function]}
                                    onExited={[Function]}
                                    onExiting={[Function]}
                                    role="alert"
                                    tag="div"
                                    timeout={150}
                                    unmountOnExit={true}
                                  >
                                    <Transition
                                      appear={true}
                                      enter={true}
                                      exit={true}
                                      in={false}
                                      mountOnEnter={false}
                                      onEnter={[Function]}
                                      onEntered={[Function]}
                                      onEntering={[Function]}
                                      onExit={[Function]}
                                      onExited={[Function]}
                                      onExiting={[Function]}
                                      timeout={150}
                                      unmountOnExit={true}
                                    />
                                  </Fade>
                                </Alert>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Email or Phone Number
                                  </label>
                                </Label>
                                <Input
                                  autoFocus={true}
                                  invalid={false}
                                  name="username"
                                  onChange={[Function]}
                                  required={true}
                                  type="text"
                                  value=""
                                >
                                  <input
                                    autoFocus={true}
                                    className="form-control"
                                    name="username"
                                    onChange={[Function]}
                                    required={true}
                                    type="text"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <FormGroup
                              tag="div"
                            >
                              <div
                                className="form-group"
                              >
                                <Label
                                  tag="label"
                                  widths={
                                    Array [
                                      "xs",
                                      "sm",
                                      "md",
                                      "lg",
                                      "xl",
                                    ]
                                  }
                                >
                                  <label
                                    className=""
                                  >
                                    Password
                                  </label>
                                </Label>
                                <Input
                                  invalid={false}
                                  name="password"
                                  onChange={[Function]}
                                  required={true}
                                  type="password"
                                  value=""
                                >
                                  <input
                                    className="form-control"
                                    name="password"
                                    onChange={[Function]}
                                    required={true}
                                    type="password"
                                    value=""
                                  />
                                </Input>
                                <FormText
                                  color="danger"
                                  tag="small"
                                >
                                  <small
                                    className="form-text text-danger"
                                  />
                                </FormText>
                              </div>
                            </FormGroup>
                            <Button
                              block={true}
                              color="primary"
                              tag="button"
                            >
                              <button
                                aria-label={null}
                                className="btn btn-primary btn-block"
                                onClick={[Function]}
                              >
                                SIGN IN
                              </button>
                            </Button>
                          </form>
                        </Form>
                      </div>
                    </Col>
                  </div>
                </Row>
              </div>
            </Container>
          </StyledComponent>
        </Styled(Container)>
      </Login>
    </Connect(Login)>
  </Provider>
</TestProvider>
`;
